hybridse/src/udf/udf.cc

```diff
@@ -64,6 +64,45 @@ void hex(StringRef *str, StringRef *output) {
     output->data_ = buffer;
 }
 
+void unhex(StringRef *str, StringRef *output) {
+    std::ostringstream ss;
+    uint8_t* arr = new uint8_t[str->size_];
+    memset(arr, 0, str->size_);
+    bool flag = 0;
+    for (uint32_t i=0; i < str->size_; i++) {
+        // only recognize 'A' to 'F' and 'a' to 'f' and '0' to '9', otherwise return null.
+        if (str->data_[i] <= 'F' && str->data_[i] >= 'A') {
+            arr[i] = str->data_[i] - 'A' + 10;
+        } else if (str->data_[i] <= 'f' && str->data_[i] >= 'a') {
+            arr[i] = str->data_[i] - 'a' + 10;
+        } else if (str->data_[i] <= '9' && str->data_[i] >= '0') {
+            arr[i] = str->data_[i] - '0';
+        } else {
+            flag = 1;
+            break;
+        }
+    }
+    if (flag == 0) {    // every character is valid hex character
+        if (str->size_ % 2 == 0) {
+            for (uint32_t i=0; i < str->size_; i+=2) {
+                ss << char(arr[i] << 4 | arr[i+1]);
+            }
+        } else {
+            ss << char(arr[0]);
+            for (uint32_t i=1; i < str->size_; i+=2) {
+                ss << char(arr[i] << 4 | arr[i+1]);
+            }
+        }
+        output->size_ = ss.str().size();
+        char *buffer = AllocManagedStringBuf(output->size_);
+        memcpy(buffer, ss.str().data(), output->size_);
+        output->data_ = buffer;
+    } else {
+        output->size_ = 4;
+        output->data_ = "NULL";
```

Comment:
it is the `NULL` value as stated above

Reply:
according to spark sql, when I input the invalid characters, it will return the following:
```shell
spark-sql> select unhex("zzz");
NULL
Time taken: 2.967 seconds, Fetched 1 row(s)
spark-sql> select char_length(unhex("zzz"));
NULL
Time taken: 0.087 seconds, Fetched 1 row(s)
```
So in my code, I write the following:
```cpp
output->size_ = NULL;
output->data_ = NULL;
```
in `udf_ir_builder_test.cc` file, 
```cpp
CheckUdf<StringRef, StringRef>("unhex", nullptr, StringRef("Zzzz"));
```
And I can pass all the unit tests.
Is that all right?

Reply:
in this case (invalid input), set the third parameter (is_bool) to true is the only step, resulting the output as `NULL`. Generally the `output` object here is not accessed afterwards.